pkg/reconciler/pipelinerun/cancel.go

```diff
@@ -80,17 +114,21 @@ func cancelPipelineRun(ctx context.Context, logger *zap.SugaredLogger, pr *v1bet
 			continue
 		}
 	}
-	// If we successfully cancelled all the TaskRuns and Runs, we can consider the PipelineRun cancelled.
-	if len(errs) == 0 {
-		pr.Status.SetCondition(&apis.Condition{
-			Type:    apis.ConditionSucceeded,
-			Status:  corev1.ConditionFalse,
-			Reason:  ReasonCancelled,
-			Message: fmt.Sprintf("PipelineRun %q was cancelled", pr.Name),
-		})
-		// update pr completed time
-		pr.Status.CompletionTime = &metav1.Time{Time: time.Now()}
-	} else {
+
+	return errs
+}
+
+// gracefullyCancelPipelineRun marks any non-final resolved TaskRun(s) as cancelled and runs finally.
+func gracefullyCancelPipelineRun(ctx context.Context, logger *zap.SugaredLogger, pr *v1beta1.PipelineRun, pipelineRunFacts *resources.PipelineRunFacts, clientSet clientset.Interface) error {
+	if !pipelineRunFacts.IsRunning() {
+		// There is no running DAG task to be cancelled. Skipping this action if already running finally.
+		return nil
+	}
+
+	errs := cancelPipelineTaskRuns(ctx, logger, pr, clientSet)
```

Comment:
Also, wouldn't this now cancel the running finally tasks? 🤔 earlier we had a check here to see if a taskRun belongs to `finally` dag and if so, we did not cancel that particular taskRun.

Reply:
sorry for confusion, I see how you changed this now. Do not attempt to cancel if any of the `dag` is not running, i.e. any of the finally is running. I think it might be better to check explicitly if any of the finally is running i.e `return nil if isFinallyRunning()` because `isRunning` returns `false` if there is no `taskRun` is associated with a `dag`.  It could be possible that nothing is scheduled yet and pipeline is just starting.

Please feel free to explain/counter it if I am not understanding it right.

Reply:
In case the has been no taskrun created yet `isRunning` will return false, so `gracefullyCancelPipelineRun` will not be executed, and that will continue to happen once final tasks are running, so unless I'm missing something, I think this specific part of the code is ok. @pritidesai wdyt?
@pritidesai comment made me thinking at the case where there is no taskrun in the state, and there are three possible cases that I can think can get us to that situation:

1. the pipeline is just starting - no taskrun created yet. In this case the graceful cancel method will not cancel anything, because there is nothing to cancel, which is ok. No new task from the DAG will be scheduled as this is checked for in the DAGExecutionQueue method, which is ok. However, no new task from finally will be scheduled either because GetFinalTask will only check whether all Tasks are done or at least one is failed, which won’t match is no task was cancelled. So I think this looks like an issue.
2. the pipeline is in pending mode: there is no check for a graceful cancel before the pending check, which means that a graceful cancel will be ignored on a pipeline in pending mode. I think this might be considered ok if there is a finally section to run, because the pipeline shouldn’t run anything (even finally) until it’s pending mode, but if there is no finally section it could make sense to cancel the pipeline. This is not a big deal, but it might be worth documenting that a pipeline in pending mode will need a forced cancel to be cancelled.
3. the pipeline was cancelled before any taskrun could be started, and it’s now running the finally section (this may only happen once the first issue is solved). For this one, once the first issue is solved, it should behave correctly

Reply:
As @rafalbigaj correctly pointed out, `checkTasksDone` will return true is the DAG state is empty, so (1) - and thus (3) will work correctly today. I think we should document (2), but I'd be happy for it to be a follow up docs PR, as long as it happens in the same release.

Reply:
One thing that we should fix in the future though is the behaviour of `checkDAGTasksDone`  and `checkFinalTasksDone`.  The current code reads:

```
// check if all DAG tasks done executing (succeeded, failed, or skipped)
func (facts *PipelineRunFacts) checkDAGTasksDone() bool {
	return facts.checkTasksDone(facts.TasksGraph)
}
```

We assume that if tasks are not in the state they were skipped, but I don't think that's correct?
This works todays anyways because of the way those functions are used, but I think we should fix the logic or at least the comments to match the actual behaviour.

Reply:
If I'm not mistaken, the behaviour with no TaskRun is verified by `TestReconcileOnCancelledRunFinallyPipelineRunWithFinalTask`

Reply:
@afrittoli yes, that's correct

Reply:
@pritidesai My understanding is that there is no need to cancel dag tasks if none is running. Therefore, `IsRunning` check seems to me to be quite natural.